rfd/0151-database-permission-management.md

```diff
@@ -0,0 +1,421 @@
+
+All of the above fields are optional, and more may be added in the future if
+needed. The database-specific implementation is responsible for populating these
+fields from the database schema.
+
+### Database object permissions
+
+The permissions for particular objects are defined in a role using the new
+`db_permissions` field, found under `spec.allow` and `spec.deny` respectively.
+Each permission specifies:
+
+- A list of labels the database object must match.
+- A list of permissions that will be given to the object for the user.
+
+As an example, the role `db_read_non_confidential` allows access to tables and
+views with a label `confidential: false` and explicitly disallows any access to
+those labeled `confidential: true`.
+
+```yaml
+kind: role
+version: v7
+metadata:
+  name: db_read_non_confidential
+spec:
+  allow:
+    db_permissions:
+      - match:
+          confidential: 'false'
+          kind:
+            - table
+            - view
+        permissions:
+          - SELECT
+          - INSERT
+          - UPDATE
+  deny:
+    db_permissions:
+      - match:
+          confidential: 'true'
+        permissions:
+          - '*'
+```
+
+The object attributes are not replicated automatically as labels. In the example
+above, `kind` is a label which must be applied by an import rule, for example
+one as follows:
+
+```yaml
+kind: db_object_import_rule
+version: v1
+metadata:
+  name: import_object_kind
+spec:
+  mappings:
+    - add_labels:
+        kind: table
+      match:
+        object_kinds:
+          - table
+    - add_labels:
+        kind: view
+      match:
+        object_kinds:
+          - view
+    - add_labels:
+        kind: schema
+      match:
+        object_kinds:
+          - schema
+  priority: 100
+```
+
+#### Applying permissions
+
+The permissions will be applied to the user after the user is provisioned in the
+database. The exact mechanism will be database-specific; for example, in SQL
+databases like Postgres or MySQL this will be done through appropriate `GRANT`
+statements, executed through a helper stored procedure.
+
+After the session is finished, the user is removed/deactivated, and _all_
+permissions must be revoked. Again, for SQL databases, a corresponding `REVOKE`
+statements will be issued. To ensure complete removal of all permissions, the
+stored procedure will iterate over all schemas and objects within, revoking the
+access to each individual object.
+
+Additionally, the permissions _may_ be updated if a schema change is detected.
+
+The permissions also _may_ be updated if the role definition in the backend
+changes, but the cost of implementing such a feature should be weighed against
+the benefits it would provide. The list of roles for a given user is unchanging
+in the scope of a single connection, so this is not an element that may change.
+
+To avoid confusion regarding the source of access, `db_permissions` will only be
+considered if no `db_roles` are configured for user.
```

Comment:
Nit: Clarify that these two settings are mutually exclusive.

Reply:
I've rephrased this sentence to match.